doc/source/design/design-adapt.rst

```diff
@@ -0,0 +1,273 @@
+* :math:`\hat{L}_i^{k+1}` *block i's desired next level as locally-evaluated from refinement criteria*
+* :math:`\underline{L}_{i,s}^{k+1} \leq L_i^{k+1} \leq
+  \bar{L}_{i,s}^{k+1}`: *current lower and upper level bounds (for
+  step s), which are dynamically updated*
+* :math:`L_i^{k+1}` *the next level which is decided when* :math:`\underline{L}_{i,s}^{k+1} = \bar{L}_{i,s}^{k+1}`
+ 
+We can now write the two main conditions that we use to initialize and
+update the level bounds:
+
+* :math:`|L_i^k - L_i^{k+1}| \le 1` *the (temporal) level-jump condition: a block can refine or coarsen at most once per adapt cycle*
+* :math:`|L_i^{k} - L_j^{k}| \le 1` *the  (spacial) level-jump condition: refinement levels of adjacent blocks can differ by at most one*
+
+Level bounds are initialized to be :math:`\underline{L}_{i,0}^{k+1}
+\leftarrow \hat{L}_i^{k+1}` and :math:`\bar{L}_{i,0}^{k+1} \leftarrow
+L_i^{k} + 1`. That is, the minimum level is initially the level
+determined by the local refinement criteria, and the maximum level is
+initially one level of refinement more than the current level (or the
+maximum allowed level in the simulation.)
+
+The balancing step of the algorithm proceeds by alternately sending a
+block's level bounds to its neighbors, and updating the block's level
+bounds given received updated bounds from its neighbors. Bounds are updated
+according to the following:
+
+:math:`\underline{L}_{i,s+1}^{k+1} \leftarrow \max ( \underline{L}_{i,s}^{k+1}, \max_j (\underline{L}_{j,s}^{k+1} - 1))`
+
+:math:`\bar{L}_{i,s+1}^{k+1} \leftarrow \max ( \underline{L}_{i,s}^{k+1}, \max_j(\bar{L}_{j,s}^{k+1} - 1))`
+
+The lower bound, which acts like the current desired level in the
+previous algorithm, is updated if any neighbor's minimum bound is
+greater than one plus the block's current minimum bound.  The maximum
+bound, which is used to determine when the balancing algorithm
+terminates, is defined as the maximum of the minimum bound, and the
+maximum of `all` neighboring maximum bounds minus one. Note that in
+general the maximum bound can only be updated after all neighboring
+blocks have been heard from. Additional synchronization is required
+for a block to coarsen, since a block can coarsen only if all of its siblings
+can as well.
+
+======================================
+Revised adapt algorithm implementation
+======================================
+
+To reduce the complexity of the already over-burdened Block classes,
+we introduce an ``Adapt`` class to maintain and update level bounds
+for a Block and its neighbors. The ``Adapt`` class keeps track of the
+current level bounds of all neighboring blocks, which is redundantly
+stored as a list of ``LevelInfo`` objects for each neighboring Block,
+and a ``face_level_`` vector of the current level in the direction of
+each face (represented as a tuple of values -1, 0, or 1, and is
+accessed e.g. during flux-correction). Below summarizes the API for
```

Comment:
The way that ``LevelInfo`` works is pretty self-evident from the class definition. But the way that ``face_level_`` works is much less evident. Would you mind elaborating more on how these encode the same information?

Reply:
I don't want to get too technical in the documentation. How about:

> (The "face_level_" representation is a carry-over from the
> previous algorithm, but was retained because it simplifies code that needs to access a
> neighbor's level given the neighbor's relative direction rather than absolute Index).